File: bindings/java/samples/Sample_mips.java
/*

Java bindings for the Unicorn Emulator Engine

Copyright(c) 2015 Chris Eagle

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/

/* Unicorn Emulator Engine */
/* By Nguyen Anh Quynh, 2015 */

/* Sample code to demonstrate how to emulate Mips code (big endian) */

package samples;

import unicorn.*;

public class Sample_mips implements UnicornConst, MipsConst {

    // code to be emulated
    public static final byte[] MIPS_CODE_EB = { 52, 33, 52, 86 }; // ori $at, $at, 0x3456
    public static final byte[] MIPS_CODE_EL = { 86, 52, 33, 52 }; // ori $at, $at, 0x3456

    // memory address where emulation starts
    public static final int ADDRESS = 0x10000;

    // callback for tracing basic blocks
    private static class MyBlockHook implements BlockHook {
        public void hook(Unicorn u, long address, int size, Object user_data) {
            System.out.format(
                ">>> Tracing basic block at 0x%x, block size = 0x%x\n", address,
                size);
        }
    }

    // callback for tracing instruction
    private static class MyCodeHook implements CodeHook {
        public void hook(Unicorn u, long address, int size, Object user_data) {
            System.out.format(
                ">>> Tracing instruction at 0x%x, instruction size = 0x%x\n",
                address, size);
        }
    }

    public static void test_mips_eb() {

        long r1 = 0x6789L;     // R1 register

        System.out.println("Emulate MIPS code (big-endian)");

        // Initialize emulator in MIPS mode
        Unicorn u =
            new Unicorn(UC_ARCH_MIPS, UC_MODE_MIPS32 + UC_MODE_BIG_ENDIAN);

        // map 2MB memory for this emulation
        u.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        u.mem_write(ADDRESS, MIPS_CODE_EB);

        // initialize machine registers
        u.reg_write(UC_MIPS_REG_1, r1);

        // tracing all basic blocks with customized callback
        u.hook_add(new MyBlockHook(), 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        u.hook_add(new MyCodeHook(), ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        u.emu_start(ADDRESS, ADDRESS + MIPS_CODE_EB.length, 0, 0);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");

        r1 = u.reg_read(UC_MIPS_REG_1);
        System.out.format(">>> R1 = 0x%x\n", r1);
    }

    public static void test_mips_el() {
        long r1 = 0x6789L;     // R1 register

        System.out.println("Emulate MIPS code (little-endian)");

        // Initialize emulator in MIPS mode
        Unicorn u = new Unicorn(UC_ARCH_MIPS,
            UC_MODE_MIPS32 + UC_MODE_LITTLE_ENDIAN);

        // map 2MB memory for this emulation
        u.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        u.mem_write(ADDRESS, MIPS_CODE_EL);

        // initialize machine registers
        u.reg_write(UC_MIPS_REG_1, r1);

        // tracing all basic blocks with customized callback
        u.hook_add(new MyBlockHook(), 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        u.hook_add(new MyCodeHook(), ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        u.emu_start(ADDRESS, ADDRESS + MIPS_CODE_EL.length, 0, 0);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");

        r1 = u.reg_read(UC_MIPS_REG_1);
        System.out.format(">>> R1 = 0x%x\n", r1);
    }

    public static void main(String args[]) {
        test_mips_eb();
        System.out.println("===========================");
        test_mips_el();
    }
}

File: bindings/haskell/samples/SampleMips.hs
-- Sample code to demonstrate how to emulate Mips code (big endian)

import Unicorn
import Unicorn.Hook
import qualified Unicorn.CPU.Mips as Mips

import qualified Data.ByteString as BS
import Data.Word
import qualified Numeric as N (showHex)

-- Code to be emulated
--
-- ori $at, $at, 0x3456
mipsCodeEb :: BS.ByteString
mipsCodeEb = BS.pack [0x34, 0x21, 0x34, 0x56]

-- ori $at, $at, 0x3456
mipsCodeEl :: BS.ByteString
mipsCodeEl = BS.pack [0x56, 0x34, 0x21, 0x34]

-- Memory address where emulation starts
address :: Word64
address = 0x10000

-- Pretty-print integral as hex
showHex :: (Integral a, Show a) => a -> String
showHex =
    flip N.showHex ""

-- Calculate code length
codeLength :: Num a => BS.ByteString -> a
codeLength =
    fromIntegral . BS.length

hookBlock :: BlockHook ()
hookBlock _ addr size _ =
    putStrLn $ ">>> Tracing basic block at 0x" ++ showHex addr ++
               ", block size = 0x" ++ (maybe "0" showHex size)

hookCode :: CodeHook ()
hookCode _ addr size _ =
    putStrLn $ ">>> Tracing instruction at 0x" ++ showHex addr ++
               ", instruction size = 0x" ++ (maybe "0" showHex size)

testMipsEb :: IO ()
testMipsEb = do
    putStrLn "Emulate MIPS code (big-endian)"

    result <- runEmulator $ do
        -- Initialize emulator in MIPS mode
        uc <- open ArchMips [ModeMips32, ModeBigEndian]

        -- Map 2MB memory for this emulation
        memMap uc address (2 * 1024 * 1024) [ProtAll]

        -- Write machine code to be emulated to memory
        memWrite uc address mipsCodeEb

        -- Initialise machine registers
        regWrite uc Mips.Reg1 0x6789

        -- Tracing all basic blocks with customized callback
        blockHookAdd uc hookBlock () 1 0

        -- Tracing one instruction at address with customized callback
        codeHookAdd uc hookCode () address address

        -- Emulate machine code in infinite time (last param = Nothing), or
        -- when finishing all the code
        let codeLen = codeLength mipsCodeEb
        start uc address (address + codeLen) Nothing Nothing

        -- Return the results
        r1 <- regRead uc Mips.Reg1

        return r1
    case result of
        Right r1 -> do
            -- Now print out some registers
            putStrLn ">>> Emulation done. Below is the CPU context"
            putStrLn $ ">>> R1 = 0x" ++ showHex r1
        Left err -> putStrLn $ "Failed with error: " ++ show err ++ " (" ++
                               strerror err ++ ")"

testMipsEl :: IO ()
testMipsEl = do
    putStrLn "==========================="
    putStrLn "Emulate MIPS code (little-endian)"

    result <- runEmulator $ do
        -- Initialize emulator in MIPS mode
        uc <- open ArchMips [ModeMips32, ModeLittleEndian]

        -- Map 2MB memory for this emulation
        memMap uc address (2 * 1024 * 1024) [ProtAll]

        -- Write machine code to be emulated to memory
        memWrite uc address mipsCodeEl

        -- Initialize machine registers
        regWrite uc Mips.Reg1 0x6789

        -- Tracing all basic blocks with customized callback
        blockHookAdd uc hookBlock () 1 0

        -- Tracing one instruction at address with customized callback
        codeHookAdd uc hookCode () address address

        -- Emulate machine code in infinite time (last param = Nothing), or
        -- when finishing all the code
        let codeLen = codeLength mipsCodeEl
        start uc address (address + codeLen) Nothing Nothing

        -- Return the results
        r1 <- regRead uc Mips.Reg1

        return r1
    case result of
        Right r1 -> do
            -- Now print out some registers
            putStrLn ">>> Emulation done. Below is the CPU context"
            putStrLn $ ">>> R1 = 0x" ++ showHex r1
        Left err -> putStrLn $ "Failed with error: " ++ show err ++ " (" ++
                               strerror err ++ ")"

main :: IO ()
main = do
    testMipsEb
    testMipsEl
